netbench/netbench-collector/src/main.rs

```diff
@@ -18,24 +36,363 @@ pub struct Args {
+                from_location,
+                other_state
+            );
+            other_state
+        } else {
+            log_polling!(
+                verbose,
+                "{} unavailable, assuming state: {:?}",
+                from_location,
+                assume_on_no_response
+            );
+            assume_on_no_response
+        }
+    }
+    fn poll(
+        &self,
+        wait_for: State,
+        assume_on_no_response: State,
+        initial_delay: Duration,
+        poll_delay: Duration,
+    ) -> JoinHandle<io::Result<()>> {
+        let other_location = self.other_location.clone();
+        let verbose = self.verbose;
+        tokio::spawn(async move {
+            log_polling!(
+                verbose,
+                "Initial Delay Sleeping for {} seconds",
+                initial_delay.as_secs()
+            );
+            sleep(initial_delay).await; // Initial Delay
+            loop {
+                let new_state =
+                    Self::get_state(other_location.clone(), assume_on_no_response, verbose).await;
+                if new_state == wait_for {
+                    break;
+                }
+                log_polling!(verbose, "Delaying for {} seconds", poll_delay.as_secs());
+                sleep(poll_delay).await;
+            }
+            Ok(())
+        })
+    }
+    async fn serve(&self) -> Result<JoinHandle<io::Result<()>>> {
+        let listener = TcpListener::bind(self.location.as_str()).await?;
+        let current_state = self.current_state.clone();
+        let verbose = self.verbose;
+        Ok(tokio::spawn(async move {
+            loop {
+                let state: State = current_state
+                    .clone()
+                    .load(Ordering::Relaxed)
+                    .try_into()
+                    .expect("An invalid atomic u8 got constructed.");
+                if state == State::Finished {
+                    log_serving!(verbose, "We are done; stop serving state");
+                    break Err(io::Error::new(ErrorKind::Other, "Finished"));
+                }
+                let (mut socket, _) = match timeout(Duration::from_secs(5), listener.accept()).await
+                {
+                    Ok(Ok(o)) => o,
+                    _ => {
+                        log_serving!(verbose, "No peers made connection in the last 5 seconds; checking for finished status");
+                        continue;
+                    }
+                };
+                let served_state: State = current_state
+                    .load(Ordering::Relaxed)
+                    .try_into()
+                    .expect("An invalid atomic u8 got constructed.");
+                socket.write_all(&serde_json::to_vec(&StateMessage::from(served_state)).unwrap()).await?;
+                log_serving!(verbose, "Told peer we are in state {:?}", served_state);
+            }
+        }))
+    }
+}
+
+macro_rules! log_server {
```

Comment:
Tracing will define macros like 
```
info!("very useful");
debug!("less useful");
```
And then you can set the "log level" to e.g. "DEBUG" to get all of the outputs, or set the log level to "INFO" to only get the "very useful" log statement.